File: DBD_heeje/Day_51~60/D55_2493.py
# [BOJ] 2493. 탑(PyPy3)
# 소요 시간 : 10 분
# 실행 시간 : 304 ms
# 메모리 : 212324 KB

import sys
input = sys.stdin.readline

N = int(input())
towers = list(map(int, input().split()))

answer = [0] * N
stack = []

for i in range(N - 1, -1, -1):
    while stack and towers[stack[-1]] < towers[i]:
        answer[stack.pop()] = i + 1
    
    stack.append(i)

print(*answer)

File: DBD_heeje/Day_51~60/D55_2493.java
// [BOJ] 2493. 탑
// 실행 시간 : 808 ms
// 메모리 : 97776 KB

import java.util.*;
import java.io.*;

class Main {
	public static StringTokenizer st;
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		
		Stack<Integer> stackInt = new Stack<>();
		
		st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		
		st = new StringTokenizer(br.readLine());
		int[] towers = new int[N];
		int[] answer = new int[N];
		for (int i = 0; i < N; i++) {
			towers[i] = Integer.parseInt(st.nextToken());
		}
		
		for (int i = N - 1; i >= 0; i--) {
			while (!stackInt.isEmpty() && towers[stackInt.peek()] < towers[i]) {
				answer[stackInt.pop()] = i + 1;
			}
			
			stackInt.add(i);
		}
		
		for (int i = 0; i < N; i++) {
			sb.append(answer[i] + " ");
		}
		
		System.out.println(sb);
	}
}